### DISCUSSION TITLE
Complete example with ICCameraDevice::mediaFiles

### DISCUSSION THREAD

**Discussion by mikeumus** [brandonhamilton/image-capture-core-rs#6] [Ideas]
Grateful for this wonderful repo but am taking time trying to get it all the way to calling methods on the attached camera. Without this, the library can maybe give the name of the attached device, but that's about it. 

Need to see how to get all the way to `ICCameraDevice::mediaFiles(camera_id);`

That'd be great!
Please consider turning on GitHub Discussions and Sponsor 💖 

Thank you @brandonhamilton 🙇🏻‍♂️

**Comment by mikeumus** [comment]
@brandonhamilton I found this repo from @joride today:
https://github.com/Joride/Monarch

@joride, you really went all out on that implementation. I'm in awe. Bravo! 👏 
You haven't been wondering about trying out Rust have you? 😇

**Reply by Joride** [reply]
Thanks for your kind words mikeumus, happy to hear it helps you out! 
No rust for me at the moment, I have my plate full with some other things 😃.

**Reply by mikeumus** [reply]
It's a great language and ecosystem! Keep it in mind for the future! 🌄

**Comment by mikeumus** [comment]
I've almost have an open session where I should be able to access `ICCameraDevice::mediaFiles()` but am a little stuck and wanted to shoot out this 🆘 before going to sleep again:

I can create an `ICCameraDevice` now and assign `ICDevice` a delegate, but after calling `device.requestOpenSession()`, `device.hasOpenSession()` never turns 1 (true),see below:


```rust
unsafe {
    let camera_device: &dyn ICCameraDevice = &device.clone();
    ICDevice::requestOpenSession(device);
    let requestOpen = device.requestOpenSession();
    if let requestOpen = () {
        println!("  📸📸 image_capture_core_mod hasOpenSession: {:?}", ICDevice::hasOpenSession(device));
        println!("  📸📸 image_capture_core_mod hasOpenSession: {:?}", device.hasOpenSession());
    }
    std::thread::sleep(std::time::Duration::from_secs(7));
    println!("  📸📸 image_capture_core_mod hasOpenSession: {:?}", device.hasOpenSession());
    std::thread::sleep(std::time::Duration::from_secs(7));
    println!("  📸📸 image_capture_core_mod hasOpenSession: {:?}", ICDevice::hasOpenSession(device));
    std::thread::sleep(std::time::Duration::from_secs(7));
    println!("  📸📸 image_capture_core_mod hasOpenSession: {:?}", device.hasOpenSession());
    
    println!("  📸📸 image_capture_core_mod device.delegate: {:?}", ICDevice::delegate(device));
}
```

_ _ _

Do you think it might only work if we implement this version of `requestOpenSession(options:completion:)` with the completion function?
- https://developer.apple.com/documentation/imagecapturecore/icdevice/3142916-requestopensession

Or do you think I may be going at it wrong? 
I'd appreciate any sight you may be able to provide @brandonhamilton 🙇🏻‍♂️
Were you ever able to access `contents` or `mediaFiles` before with the lib as is? 

I'll update the readme and any PR fixes once `mediaFiles()` or `contents` starts showing up. 🙂👍

**Comment by mikeumus** [comment]
An Update*

I have access to the camera session and `mediaFiles` now. 
Just working on turning the `Object` it returns into a `Vec<String>` or something close to that. 
Stopping for tonight. 😴

**Comment by mikeumus** [comment]
This is where I'm leaving off tonight, trying to convert `mediaFiles` or `contents` into `Vec<ICCameraItem>`

```rust
    extern "C" fn device_did_open_session_with_error(
        obj_obj: &Object,
        sel_sel: Sel,
        device: id,
        what_dis: id,
    ) {
        println!("  📸📸 device_did_open_session_with_error obj: {:?}", obj_obj);
        println!("  📸📸 device_did_open_session_with_error sel: {:?}", sel_sel);
        println!("  📸📸 device_did_open_session_with_error what_Dir: {:?}", what_dis);

        unsafe {
            let cam_items: Vec<&dyn ICCameraItem> = device
                .mediaFiles().as_ref().iter()
                .map(|cam_item| <&dyn ICCameraItem>::into(cam_item) )
                .collect();
        }
}
```

Getting this error:
``` bash
error[E0277]: the trait bound `&Object: ICCameraItem` is not satisfied
  --> src/image_capture_core_mod.rs:86:59
   |
86 |                 .map(|cam_item| <&dyn ICCameraItem>::into(cam_item) )
   |                                                           ^^^^^^^^ the trait `ICCameraItem` is not implemented for `&Object`
   |
   = note: required for the cast to the object type `dyn ICCameraItem`
   ```
   
   Which makes me think I probably need a `ICCameraItem::new()` type function like Joride has here:
https://github.com/Joride/Monarch/blob/main/Shared/CameraDeviceManager.swift#L269

back to sleep for now... 😴
